src/memory/spi_mem.c

```diff
@@ -0,0 +1,600 @@
+    SPI_MEM_exchange_block(buffer, 1);
+    _spi_mem_cs_high();
+
+    // --- Chip Erase ---
+    buffer[0] = CMD_CE;
+    _spi_mem_cs_low();
+    SPI_MEM_exchange_block(buffer, 1);
+    _spi_mem_cs_high();
+
+    _spi_mem_wait();
+    util_log("Chip erased");
+}
+
+bool spi_mem_sector_erase(uint32_t sector_addr)
+{
+    if (sector_addr & ~SECTOR_MASK || (sector_addr + SECTOR_SIZE - 1) > MEMORY_LIMIT) {
+        util_log("Invalid sector address %p", (void*)(uintptr_t)sector_addr);
+        return false;
+    }
+
+    uint8_t buffer[PAGE_SIZE + 4];
+    // --- Enable Write ---
+    buffer[0] = CMD_WREN;
+    _spi_mem_cs_low();
+    SPI_MEM_exchange_block(buffer, 1);
+    _spi_mem_cs_high();
+
+    // --- Sector Erase (write 4 bytes) ---
+    buffer[0] = CMD_SE;
+    buffer[1] = (sector_addr >> 16) & 0xFF;
+    buffer[2] = (sector_addr >> 8) & 0xFF;
+    buffer[3] = sector_addr & 0xFF;
+
+    _spi_mem_cs_low();
+    SPI_MEM_exchange_block(buffer, 4);
+    _spi_mem_cs_high();
+
+    // --- Wait for write to end ---
+    _spi_mem_wait();
+
+    return true;
+}
+
+bool spi_mem_page_read(uint32_t page_addr, uint8_t* data_out)
+{
+    if (page_addr % PAGE_SIZE != 0) {
+        util_log("Invalid page read address %p", (void*)(uintptr_t)page_addr);
+        return false;
+    }
+
+    uint8_t tmp_buf[PAGE_SIZE + 4];
+    _spi_mem_read(page_addr, PAGE_SIZE, tmp_buf);
+    memcpy(data_out, &tmp_buf[4], PAGE_SIZE);
+    return true;
+}
+
+uint8_t* spi_mem_read(uint32_t address, size_t size)
+{
+    if (address + size - 1 > MEMORY_LIMIT || size < 1) {
+        util_log("Invalid read address %p or size %i", (void*)(uintptr_t)address, (int)size);
+        return NULL;
+    }
+
+    uint8_t* buffer = malloc(size + 4);
+    if (!buffer) {
+        util_log("Memory allocation failed");
+        return NULL;
+    }
+
+    _spi_mem_read(address, size, buffer);
+
+    // shift the read data at the beginning of the buffer, overriding the command and the address.
+    for (int i = 0; i < (int)size; i++) {
+        buffer[i] = buffer[i + 4];
+    }
+    return buffer;
+}
+
+bool spi_mem_page_write(uint32_t page_addr, const uint8_t* input)
+{
+    if (page_addr % PAGE_SIZE != 0) {
+        util_log("Invalid page write address %p", (void*)(uintptr_t)page_addr);
+        return false;
+    }
+
+    uint8_t buffer[PAGE_SIZE + 4];
+    // --- Enable Write ---
+    buffer[0] = CMD_WREN;
+    _spi_mem_cs_low();
+    SPI_MEM_exchange_block(buffer, 1);
+    _spi_mem_cs_high();
+
+    // --- Page Program (write 4 bytes) ---
+    buffer[0] = CMD_PP;
+    buffer[1] = (page_addr >> 16) & 0xFF;
+    buffer[2] = (page_addr >> 8) & 0xFF;
+    buffer[3] = page_addr & 0xFF;
+    memcpy(&buffer[4], input, PAGE_SIZE);
+
+    _spi_mem_cs_low();
+    SPI_MEM_exchange_block(buffer, 4 + PAGE_SIZE);
+    _spi_mem_cs_high();
+
+    // --- Wait for write to end ---
+    _spi_mem_wait();
+
+    return true;
+}
+
+bool spi_mem_write(uint32_t address, size_t size, const uint8_t* input)
```

Comment:
Normally the size param comes after the data param. Would be nice to flip the order for consistency.